Clarify and add MITM Techniques (#3184)
OWASP/owasp-mastg: Document/0x04f-Testing-Network-Communication.md

Finally, verify that the server or termination proxy at which the HTTPS connection terminates is configured according to best practices. See also the [OWASP Transport Layer Protection cheat sheet](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Transport_Layer_Protection_Cheat_Sheet.md "Transport Layer Protection Cheat Sheet") and the [Qualys SSL/TLS Deployment Best Practices](https://github.com/ssllabs/research/wiki/SSL-and-TLS-Deployment-Best-Practices "Qualys SSL/TLS Deployment Best Practices").

> ## Intercepting Network Traffic

Author: ```suggestion
## Intercepting Network Traffic Through MITM
```

> Intercepting mobile app traffic is essential for security testing, allowing testers to analyze and manipulate network communications. The appropriate method depends on the app's security mechanisms and the data being transmitted.

Author: ```suggestion
Intercepting mobile app traffic is a critical aspect of security testing, enabling testers, analysts, or penetration testers to analyze and manipulate network communications to identify vulnerabilities. A key technique in this process is the **Machine-in-the-Middle (MITM)** attack (also known as "Man-in-the-Middle," "Manipulator-in-the-Middle," "Adversary-in-the-Middle," etc.), where the _attacker_ positions their machine between two communicating entities, typically the mobile app (client) and the servers it is communicating with. By doing so, the attacker's machine intercepts and monitors the data being transmitted between the different parties.

This technique is twofold:

- Typically **used by malicious attackers** to intercept, monitor, and potentially alter the communication without either party (app or server) being aware. This allows for malicious activities such as eavesdropping, injecting malicious content, or manipulating the data being exchanged.
- However, **in the context of the OWASP MASTG** and mobile app security testing, we use it as part of our techniques to allow the app tester to review, analyze, or modify the traffic to identify vulnerabilities such as unencrypted communication or weak security controls.

The specific interception method used depends on the app's security mechanisms and the nature of the data being transmitted. Each approach varies in complexity and effectiveness, depending on factors such as encryption and the app's ability to resist interference.
```

Here's an overview of interception techniques at different network layers:

| **Interception Technique** | **Example Tools** | **Note** |
|---------------------------|-------------------|-------------------|
| API hooking (`HttpUrlConnection`, `NSURLSession`, `WebRequest`) | Frida | Modifies how apps handle network requests. |
| Hooking TLS functions (`SSL_read`, `SSL_write`) | Frida, SSL Kill Switch | Intercepts encrypted data before it reaches the app. |
| Proxy interception | Burp Suite, ZAP, mitmproxy | Requires app to respect proxy settings. |
| Packet sniffing | `tcpdump`, Wireshark | Captures **all** TCP/UDP traffic but does **not** decrypt HTTPS. |
| MITM via ARP spoofing | bettercap | Tricks devices into sending their traffic through the attacker's machine even when the network isn't controlled by the attacker. |
| Rogue Wi-Fi AP | `hostapd`, `dnsmasq`, `iptables`, `wpa_supplicant`, `airmon-ng` | Uses an access point fully controlled by the attacker. |

You can find more information on these techniques in their corresponding technique pages:

- @MASTG-TECH-0119
- @MASTG-TECH-0120
- @MASTG-TECH-0121
- @MASTG-TECH-0122
- @MASTG-TECH-0123

> - @MASTG-TECH-0124

Reviewer: ```suggestion
- @MASTG-TECH-0124 
- @MASTG-TECH-0125
```

Here's an overview of interception techniques at different network layers:

| **Interception Technique** | **Example Tools** | **Note** |
|---------------------------|-------------------|-------------------|
| API hooking (`HttpUrlConnection`, `NSURLSession`, `WebRequest`) | Frida | Modifies how apps handle network requests. |
| Hooking TLS functions (`SSL_read`, `SSL_write`) | Frida, SSL Kill Switch | Intercepts encrypted data before it reaches the app. |
| Proxy interception | Burp Suite, ZAP, mitmproxy | Requires app to respect proxy settings. |
| Packet sniffing | `tcpdump`, Wireshark | Captures **all** TCP/UDP traffic but does **not** decrypt HTTPS. |
| MITM via ARP spoofing | bettercap | Tricks devices into sending their traffic through the attacker's machine even when the network isn't controlled by the attacker. |
| Rogue Wi-Fi AP | `hostapd`, `dnsmasq`, `iptables`, `wpa_supplicant`, `airmon-ng` | Uses an access point fully controlled by the attacker. |

You can find more information on these techniques in their corresponding technique pages:

- @MASTG-TECH-0119
- @MASTG-TECH-0120
- @MASTG-TECH-0121
- @MASTG-TECH-0122
- @MASTG-TECH-0123

> - @MASTG-TECH-0124 

Author: ```suggestion
- @MASTG-TECH-0124
```

> - @MASTG-TECH-0125

Author: ```suggestion
```

| **Layer 2 (Data Link)** | Rogue Wi-Fi AP | `hostapd`, `dnsmasq`, `iptables`, `wpa_supplicant`, `airmon-ng` | Captures **all** traffic from a connected device. |

## MASTG-TECH: Intercepting HTTP Traffic Using an Interception Proxy

Interception proxies are the most common method for intercepting mobile app traffic. They work by setting up a proxy server that intercepts and logs all HTTP/HTTPS traffic between the mobile app and the server. This allows you to view and modify the requests and responses in real-time.

Several free and commercial proxy tools are available. For example: @MASTG-TOOL-0077 and @MASTG-TOOL-0079.

### Rerouting Traffic to the Proxy

To use the interception proxy, you'll need to run it on your host computer and configure the mobile app to route HTTP(S) requests to your proxy. In most cases, it is enough to set a system-wide proxy in the network settings of the mobile device - if the app uses standard HTTP APIs or popular libraries such as `okhttp`, it will automatically use the system settings.

<img src="Images/Chapters/0x04f/BURP.png" width="100%" />

### Installing the Proxy Certificate

> Using an interception proxy breaks SSL certificate verification and the app will usually fail to initiate TLS connections. So, interception proxies require you to install a custom CA certificate on the mobile device, which allows the proxy to decrypt and inspect the encrypted HTTPS traffic. However, some apps implement certificate pinning to prevent this, which requires additional steps to bypass.

Reviewer 2: ```suggestion
Using an interception proxy breaks SSL certificate verification and the app will usually fail to initiate TLS connections. Because of this, interception proxies require you to install a custom CA certificate on the mobile device, which allows the proxy to decrypt and inspect the encrypted HTTPS traffic. Depending on the platform, the installed certificate may or may not be automatically trusted by the application. Additionally,  some apps implement certificate pinning, which requires additional effort to bypass.
```

### Per-Platform Instructions

- Android: see @MASTG-TECH-0011
- iOS: see @MASTG-TECH-0063

## MASTG-TECH: Intercepting Non-HTTP Traffic Using an Interception Proxy

> Interception proxies such as @MASTG-TOOL-0077 and @MASTG-TOOL-0079 won't show non-HTTP traffic, because they aren't capable of decoding it properly by default. There are, however, Burp plugins available such as:

Reviewer 2: ```suggestion
Interception proxies such as @MASTG-TOOL-0077 and @MASTG-TOOL-0079 won't show non-HTTP traffic, because they aren't capable of decoding it properly by default. They can, however, be extended using the following tools, allowing you to intercept and manipulate non-HTTP traffic:
```

- [Burp-non-HTTP-Extension](https://github.com/summitt/Burp-Non-HTTP-Extension "Burp-non-HTTP-Extension") and
- [Mitm-relay](https://github.com/jrmdev/mitm_relay "Mitm-relay").

> These plugins can visualize non-HTTP protocols, allowing you to intercept and manipulate the traffic.

Reviewer 2: ```suggestion
```
Deleting this due to my modification above.

Note that this setup can sometimes become very tedious and is not as straightforward as testing HTTP.

## MASTG-TECH: Intercepting HTTP Traffic by Hooking Network APIs at the Application Layer

Depending on your goal while testing the app, sometimes it is enough to monitor the traffic before it reaches the network layer or when the responses are received in the app.

> This means that you don't need to deploy a fully fledged MITM attack (including ARP Spoofing attacks, etc.) if you simply want to know if a certain piece of sensitive data is being transmitted to the network. In this case, you wouldn't even have to bypass pinning, if implemented.

Reviewer 2: ```suggestion
This means that you don't need to deploy a fully fledged MITM attack (including ARP Spoofing attacks, etc.) if you simply want to determine if certain sensitive data is being transmitted to the network. With this approach, you will not interfere with any TLS verification or pinning.
```
Can we maybe rephrase this section to a general 'Pro' and 'Con' ? And we can do that for each technique then. Or 'Advantages' and 'Limitations' ? The content itself is already there, but that gives us nicer consistency and the ability to quickly compare.

- It might be very time consuming to craft the right hooking scripts to re-assemble HTTP response pairs (across many method calls and execution threads). You might find [ready-made scripts](https://github.com/fanxs-t/Android-SSL_read-write-Hook/blob/master/frida-hook.py) and even for [alternative network stacks](https://codeshare.frida.re/@owen800q/okhttp3-interceptor/), but depending on the app and the platform, these scripts might need a lot of maintenance and might not _always work_.

See some examples:

- ["Universal interception. How to bypass SSL Pinning and monitor traffic of any application"](https://hackmag.com/security/ssl-sniffing/), sections "Grabbing payload prior to transmission" and "Grabbing payload prior to encryption"
- ["Frida as an Alternative to Network Tracing"](https://gaiaslastlaugh.medium.com/frida-as-an-alternative-to-network-tracing-5173cfbd7a0b)

## MASTG-TECH: Passive Eavesdropping

> This method involves passively capturing network traffic using tools such as Wireshark or tcpdump (@MASTG-TOOL-0081, @MASTG-TOOL-0080, or @MASTG-TOOL-0075). It is useful for identifying network endpoints, analyzing protocol metadata, and understanding how an app communicates with its server. However, it cannot automatically decrypt TLS-encrypted communication. That said, [TLS decryption is possible](https://wiki.wireshark.org/TLS#tls-decryption) if you can [obtain the pre-master secret](https://wiki.wireshark.org/TLS#using-the-pre-master-secret). For an example specific to Android, see [this article](https://nibarius.github.io/learning-frida/2022/05/21/sniffing-tls-traffic).

Reviewer 2: ```suggestion
This method involves passively capturing network traffic using tools such as @MASTG-TOOL-0081,  @MASTG-TOOL-0080, or @MASTG-TOOL-0075. It is useful for identifying network endpoints, analyzing protocol metadata, and understanding how an app communicates with its server. However, it cannot automatically decrypt TLS-encrypted communication. That said, [TLS decryption is possible](https://wiki.wireshark.org/TLS#tls-decryption) if you can [obtain the pre-master secret](https://wiki.wireshark.org/TLS#using-the-pre-master-secret). For an example specific to Android, see [this article](https://nibarius.github.io/learning-frida/2022/05/21/sniffing-tls-traffic).
```

> ### When Is it Useful?

Reviewer 2: ```suggestion
### When Is it Useful?
```
I would say it's useful when:
- Troubleshooting issues with active MitM'ing (identifying TLS errors, examining routing information, understanding protocol usage)
- Analyzing plaintext non-http traffic
- Analyzing encrypted traffic for side-channel attacks

For the others, there are better solutions.

Author: Thanks, @TheDauntless, could you please post a change suggestion for the text below?

Passive eavesdropping is particularly useful in the following scenarios:

- When mobile apps use cross-platform frameworks such as [Xamarin](https://www.xamarin.com/platform "Xamarin") that ignore system proxy settings.
- When mobile apps detect and prevent the use of system proxies.
- When intercepting push notifications, such as Google Cloud Messaging (GCM)/Firebase Cloud Messaging (FCM) on Android.
- When monitoring XMPP or other non-HTTP-based communication protocols.

### How Does It Work?

> Route the traffic from the mobile device to the host computer:

Reviewer 2: Passive collection is possible on a rooted Android and iOS device with tcpdump (so no need for host computer).

1. Configuring the host computer as a network gateway using built-in internet sharing options on Windows, macOS, and Linux.
2. Configure the mobile device to route HTTP(S) requests to your proxy.

>      - **Android (see @MASTG-TECH-0010)**: Use adb to forward traffic to your host computer.  

Reviewer 2: This wouldn't do anything without also actually configuring the system proxy. To be sure you have all traffic on Android, you'll have to use VPN or a Wifi hotspot

     - **iOS (see @MASTG-TECH-0062)**: On macOS, create a "Remote Virtual Interface" to capture all traffic from an iOS device.

> Once traffic is routed through your host, use tools like Wireshark or tcpdump to inspect and analyze it.

Reviewer 2: ```suggestion
Once traffic is routed through your host, use tools like @MASTG-TOOL-0081 or @MASTG-TOOL-0081 to inspect and analyze it.
```

## MASTG-TECH: Achieving a MITM Position via ARP Spoofing

> When proxy-based interception fails due to non-HTTP protocols or proxy-aware apps, **ARP Spoofing** can be used to redirect network traffic. ARP Spoofing is a **Layer 2 attack** that allows an attacker to impersonate the network gateway, forcing the mobile device to send its traffic through the attacker's machine.

Reviewer 2: ```suggestion
When proxy-based interception fails due to non-HTTP protocols or non-proxy-aware apps, **ARP Spoofing** can be used to redirect network traffic. ARP Spoofing is a **Layer 2 attack** that allows an attacker to impersonate the network gateway, forcing the mobile device to send its traffic through the attacker's machine.
```
(I actually never use ARP spoofing, as the alternatives are much nicer to me)

Author: I agree! And with these changes we clarify that and lead people to the right and most useful techniques.

This technique works against any device and operating system as the attack is executed on OSI Layer 2. When you are MITM, you might not be able to see clear text data, as the data in transit might be encrypted by TLS, but it will give you valuable information about the hosts involved, the protocols used, and the ports the app is communicating with.

> To execute an ARP Spoofing attack, you can use @MASTG-TOOL-0076 (bettercap) or set up a rogue access point that routes traffic through your host computer.

Reviewer 2: ```suggestion
To execute an ARP Spoofing attack, you can use @MASTG-TOOL-0076 or set up a rogue access point that routes traffic through your host computer.
```
If you're using a rogue AP, there's no point in ARP spoofing; data is already flowing through your host.

- When monitoring XMPP or other non-HTTP-based communication protocols.

### How Does It Work?

Passive eavesdropping can be performed in two ways:

1. **Directly on a rooted Android or jailbroken iOS device**  
   If the device is rooted (Android) or jailbroken (iOS), you can capture network traffic directly using `tcpdump` or similar tools, without needing a host computer. This allows you to monitor all outgoing and incoming packets in real time.

2. **By routing traffic through a host computer (works on both rooted/jailbroken and non-rooted/non-jailbroken devices)**  
   If direct packet capture on the device is not possible or preferred, you can route its network traffic to a host computer and analyze it using tools like @MASTG-TOOL-0081 or @MASTG-TOOL-0075. This method applies to **both rooted/jailbroken and non-rooted/non-jailbroken devices** and is typically achieved through:
   - **Using an interception proxy** to intercept and analyze HTTP/S traffic.
   - **Setting up a VPN-based capture** to redirect traffic through a controlled network tunnel.
   - **Performing ARP spoofing or setting up a transparent network tap** on a Wi-Fi network.

> For detailed setup instructions, refer to:

Author: ```suggestion
For detailed setup instructions, refer to:

```

### How Does It Work?

Passive eavesdropping can be performed in two ways:

1. **Directly on a rooted Android or jailbroken iOS device**  
   If the device is rooted (Android) or jailbroken (iOS), you can capture network traffic directly using `tcpdump` or similar tools, without needing a host computer. This allows you to monitor all outgoing and incoming packets in real time.

2. **By routing traffic through a host computer (works on both rooted/jailbroken and non-rooted/non-jailbroken devices)**  
   If direct packet capture on the device is not possible or preferred, you can route its network traffic to a host computer and analyze it using tools like @MASTG-TOOL-0081 or @MASTG-TOOL-0075. This method applies to **both rooted/jailbroken and non-rooted/non-jailbroken devices** and is typically achieved through:
   - **Using an interception proxy** to intercept and analyze HTTP/S traffic.
   - **Setting up a VPN-based capture** to redirect traffic through a controlled network tunnel.
   - **Performing ARP spoofing or setting up a transparent network tap** on a Wi-Fi network.

For detailed setup instructions, refer to:

- **Android:** @MASTG-TECH-0010  

> - **iOS:** @MASTG-TECH-0062  

Author: @TheDauntless please re-review

If that's the case, you are now able to see the complete network traffic that is sent and received by the mobile phone. This includes also DNS, DHCP and any other form of communication and can therefore be quite "noisy". You should therefore know how to use [DisplayFilters in Wireshark](https://wiki.wireshark.org/DisplayFilters "DisplayFilters") or know [how to filter in tcpdump](https://danielmiessler.com/study/tcpdump/#gs.OVQjKbk "A tcpdump Tutorial and Primer with Examples") to focus only on the relevant traffic for you.

### Using a Rogue Access Point

#### Network Setup

> A simple way to simulate a Machine-in-the-Middle (MITM) attack is to configure a network where all packets between the devices in scope and the target network are going through your host computer. In a mobile penetration test, this can be achieved by using an access point that the mobile devices and your host computer are connected to. Your host computer is then configured as a router and an access point.

Reviewer 2: > In a mobile penetration test, this can be achieved by using an access point that the mobile devices and your host computer are connected to. Your host computer is then configured as a router and an access point.

This feels like combining two things. Either you use an access point that both devices are connected to (typically to solve host isolation issues), or your host _is_ the access point.

> Following scenarios are possible:

Reviewer 2: ```suggestion
The following scenarios are possible:
```

If you haven't already, [set up your interception proxy](#mastg-tech-intercepting-http-traffic-using-an-interception-proxy).

**Tip:** When redirecting traffic, create specific rules for only the domains and IPs in scope to reduce noise from out-of-scope traffic.

Ensure that your interception proxy listens on the port defined in your redirection rule (`8080` in this case).

After redirecting traffic to your interception proxy, you need to forward it back to its original destination. The following steps set up redirection in @MASTG-TOOL-0077:

1. Open the **Proxy** tab and click on **Options**.
2. Select and edit your listener from the list of proxy listeners.
3. Navigate to the **Request Handling** tab and configure:
    - **Redirect to host**: Set this to the original server destination.
    - **Redirect to port**: Specify the original port.
    - Enable **"Force use of SSL"** (if HTTPS is used) and **"Support invisible proxy"**.

<img src="Images/Chapters/0x04f/burp_xamarin.png" width="100%" />

### Start Intercepting Traffic

Now, start using the app and trigger its functions. If configured correctly, HTTP messages should appear in your interception proxy.

> **Note:**  

> > When using Bettercap, enable **"Support invisible proxying"** under **Proxy Tab → Options → Edit Interface**.  

Author: ```suggestion
> When using Bettercap or DNS Spoofing, enable **"Support invisible proxying"** under **Proxy Tab → Options → Edit Interface**.  
```

Now you can connect your mobile devices to the access point.

## MASTG-TECH: Intercepting Xamarin Traffic

Xamarin is a mobile app development platform that allows developers to create [native Android](https://docs.microsoft.com/en-us/xamarin/android/get-started/ "Getting Started with Android") and [iOS apps](https://docs.microsoft.com/en-us/xamarin/ios/get-started/ "Getting Started with iOS") using Visual Studio and C#.

When testing a Xamarin app, setting the system proxy in the Device Wi-Fi settings will not capture any HTTP requests in your interception proxy. This is because Xamarin apps do not use the local proxy settings of your device. There are three ways to bypass this limitation:

> ### Option 1: Manipulating Xamarin's Network Stack Default Proxy

Reviewer 2: Some thoughts:
- Do we really want to have separate sections for cross-platform frameworks? Generally we tried to stay away from that.
- Do we want to modify source code? (vs dynamic patching)

Author: The content for patching was already there but we can remove it and leave the new content that uses Frida.

Patch the app to use a [default proxy](https://developer.xamarin.com/api/type/System.Net.WebProxy/ "System.Net.WebProxy Class") by adding the following code in the `OnCreate` or `Main` method:

```cs
WebRequest.DefaultWebProxy = new WebProxy("192.168.11.1", 8080);
```

Finally, recompile and sign the patched app.

Alternatively, use Frida to hook into the `WebRequest.DefaultWebProxy` property and dynamically set the proxy to your interception proxy.

### Option 2: Achieving a MITM Position via ARP Spoofing

Use @MASTG-TOOL-0076 to achieve a MITM position and redirect port 443 to your interception proxy running on localhost.

On macOS:

```bash
echo "
rdr pass inet proto tcp from any to any port 443 -> 127.0.0.1 port 8080
" | sudo pfctl -ef -
```

On Linux:

```bash
sudo iptables -t nat -A PREROUTING -p tcp --dport 443 -j DNAT --to-destination 127.0.0.1:8080
```

Lastly, enable **"Support invisible proxy"** in the listener settings of **@MASTG-TOOL-0007**.

### Option 3: DNS Spoofing

> If you can modify the device's DNS resolution ([DNS Spoofing](https://en.wikipedia.org/wiki/DNS_spoofing)), you can reroute the app's traffic to your proxy. For example, on a rooted Android device, you can add an entry in `/etc/hosts` mapping the app's server domain to your proxy machine’s IP. This makes the app believe that your machine is the legitimate server.

Reviewer 2: DNS spoofing is possible on non-rooted Android and non-rooted iOS by configuring the DNS server.

I would also say that /etc/hosts modification isn't really DNS spoofing

> To ensure proper interception, combine DNS spoofing with port redirection. When your machine receives the redirected connection, it will forward the traffic to the proxy. The proxy will then relay the traffic to the real server, effectively acting as a MITM (as done with @MASTG-TOOL-0076).

Reviewer 2: You mostly need port redirection for getting the traffic into the proxy, not for the data leaving the proxy. With DNS spoofing, your data will arrive on port 443 (eg) rather than 8080 (eg).

If you haven't already, [set up your interception proxy](#mastg-tech-intercepting-http-traffic-using-an-interception-proxy).

**Tip:** When redirecting traffic, create specific rules for only the domains and IPs in scope to reduce noise from out-of-scope traffic.

Ensure that your interception proxy listens on the port defined in your redirection rule (`8080` in this case).

After redirecting traffic to your interception proxy, you need to forward it back to its original destination. The following steps set up redirection in @MASTG-TOOL-0077:

1. Open the **Proxy** tab and click on **Options**.
2. Select and edit your listener from the list of proxy listeners.
3. Navigate to the **Request Handling** tab and configure:
    - **Redirect to host**: Set this to the original server destination.
    - **Redirect to port**: Specify the original port.
    - Enable **"Force use of SSL"** (if HTTPS is used) and **"Support invisible proxy"**.

<img src="Images/Chapters/0x04f/burp_xamarin.png" width="100%" />

### Start Intercepting Traffic

Now, start using the app and trigger its functions. If configured correctly, HTTP messages should appear in your interception proxy.

> **Note:**  

> > When using Bettercap, enable **"Support invisible proxying"** under **Proxy Tab → Options → Edit Interface**.  

Reviewer 2: Also necessary with DNS spoofing, because the client doesn't know that it's being intercepted.
